cdap-api-common/src/main/java/io/cdap/cdap/api/exception/ErrorDetailsProvider.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.api.exception;
+
+import javax.annotation.Nullable;
+
+/**
+ * Interface for providing error details.
+ *
+ * <p>
+ * Implementations of this interface can be used to provide more detailed error information
+ * for exceptions that occur within the code using {@link ProgramFailureException}.
+ * </p>
+ */
+public interface ErrorDetailsProvider<T> {
+
+  /**
+   * Returns a {@link RuntimeException} that wraps the given exception
+   * with more detailed information.
+   *
+   * @param e the exception to wrap
+   * @param conf configuration object
+   * @return {@link RuntimeException} that wraps the given exception
+   *         with more detailed information.
+   */
+  RuntimeException getExceptionDetails(Throwable e, @Nullable T conf);
```

Comment:
I'm not clear on why this interface is needed and why we want to pass the Configuration through it. Can you describe how it is intended to be used? Maybe include one example test plugin in the PR.

Reply:
This interface helps in cases where just generic exceptions like `IOException` are thrown from plugin but we want to generate further information from it. Like `Google APIs` through `GoogleJsonResponseException`, so `google-cloud` plugin classes can implement this interface, for example see PR: https://github.com/data-integrations/google-cloud/pull/1450

Reply:
The intent of adding `Configuration` is for delegating classes to load the delegate and redirect the call to delegates. For example, in this case `StageTrackingInputFormat`.

Reply:
Shouldn't it be the plugin doing it? Why it is part of the cdap-api contract that has to be invoked from the platform?

Reply:
> Shouldn't it be the plugin doing it? Why it is part of the cdap-api contract that has to be invoked from the platform?

Sorry, I didn't get your question exactly? Trying to give a shot at it: 

Plugin is implementing the interface and extracting more details from it and then returning `ProgramFailureException`.

`cdap-api` just defines the interface.

Reply:
This is the older PR: https://github.com/data-integrations/google-cloud/pull/1445 without this interface. In every plugin we had to do all the wrapping stuff to throw the `ProgramFailureException` but now it just needs to implement this one method.

Reply:
If I understand correctly, the idea is to try and consolidate the exception handling in a single place instead of in a bunch of places in each InputFormat/OutputFormat method?

I'm not sure that's actually what we want. Exceptions should generally be handled where they are thrown, otherwise a lot of context is lost and the logic is brittle. For example, some layer in between may wrap one of the exceptions and unknowingly break the exception handling. Also, a NOT_FOUND error in one part of the code (ex: writing to GCS) may be an error while a NOT_FOUND in another part of the code (ex: deleting a GCS object) should get ignored. Putting all the exception handling in a single place encourages broad general handling when it is often not a great solution.

It does sometimes make sense, like when we're setting up a HTTP server we register an Exception Handler class that handles very generic common logic like mapping various known exceptions into their corresponding http error codes, etc.

Reply:
We are not trying to handle any exceptions, the behavior can be different in every plugin. What the method does is just wrapping the exception in a particular platform exception after extracting information out of it.